### PR TITLE
feat(cognito): allow retrieval of UserPoolClient generated client secret

### PR DESCRIPTION
**Pull request by Dzhuneyt**
Consumers of the `cognito.UserPoolClient` construct currently have a way to retrieve useful information and pass that information to other pieces of infrastructure (e.g. Lambdas, ECS, Parameter Store, CfnOutput), e.g. the Client ID, which is useful for certain operations against the Cognito user pool that require a Cognito client.

However, if consumers decide to configure the Client, with a pre-generated (random) Client Secret for security reasons, by passing the `generateSecret: true` prop to `cognito.UserPoolClient`, most operations to the Cognito UserPool now require the Client ID AND Client Secret, otherwise they generate the `Unable to verify secret hash for client <client-id>` error (see [AWS support article](https://aws.amazon.com/premiumsupport/knowledge-center/cognito-unable-to-verify-secret-hash/) for more info).

Currently, the construct exposes no way to retrieve the Client Secret, forcing consumers to work around this by using custom resources or custom AWS SDK calls to retrieve it.

This PR addresses that, by exposing a new getter method of the `UserPoolClient` class, allowing consumers to read it on demand and pass it to other pieces of infrastructure, using code like:

```ts
const userPoolClient = new cognito.UserPoolClient(this, 'UserPoolClient', {
  userPool,
  generateSecret: true,
});

// Allows you to pass the generated secret to other pieces of infrastructure
const secret = userPoolClient.userPoolClientSecret;
```

Closes https://github.com/aws/aws-cdk/issues/7225

----

### All Submissions:

* [X] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [ ] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [ ] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-cognito/README.md
Comment (misterjoshua) [inline packages/@aws-cdk/aws-cognito/README.md:756]: Add an extra newline here - it will probably fix the broken build.

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (misterjoshua) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:430]: I like the idea of supporting this at the L2 level via `userPoolClientSecret`, however, I feel strongly that we ought to produce the custom resource if and only if the `userPoolClientSecret` is accessed. This would imply creating this custom resource in a getter function (`get userPoolClientSecret(): string { ... }`).

**Review by misterjoshua — COMMENTED** [review]
I like the idea and this is a great start, but I have a few thoughts. @pahud Could you please tag a few more community reviewers for this one?

File: packages/@aws-cdk/aws-cognito/test/user-pool-resource-server.integ.snapshot/manifest.json
Comment (misterjoshua) [inline packages/@aws-cdk/aws-cognito/test/user-pool-resource-server.integ.snapshot/manifest.json:85]: I wonder why these changed.

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (misterjoshua) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:432]: Nicely done. It should be possible to use a typed private member variable to hold this reference without the type laundering, but I don't have strong opinions here.

**Review by misterjoshua — COMMENTED** [review]
This looks really good. I have only a few small nitpicky comments.

I'd like to see more community feedback and perhaps some review by the CDK group, but otherwise, I'm happy with this changeset.

Thanks for your work so far!

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (Dzhuneyt) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts]: Cool suggestion. Refactored to implement it...

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (Dzhuneyt) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:456]: Good suggestion. Refactored to only create the Custom Resource if the property is accessed via a getter.

File: packages/@aws-cdk/aws-cognito/README.md
Comment (Dzhuneyt) [reply packages/@aws-cdk/aws-cognito/README.md:756]: PR linter now passes green.

**Comment by Dzhuneyt** [comment]
> This looks really good. I have only a few small nitpicky comments.
> 
> I'd like to see more community feedback and perhaps some review by the CDK group, but otherwise, I'm happy with this changeset.
> 
> Thanks for your work so far!

I wonder what's the best way to involve more people from the CDK group? Maybe in the official CDK Slack community?

**Comment by misterjoshua** [comment]
> > This looks really good. I have only a few small nitpicky comments.
> > I'd like to see more community feedback and perhaps some review by the CDK group, but otherwise, I'm happy with this changeset.
> > Thanks for your work so far!
> 
> I wonder what's the best way to involve more people from the CDK group? Maybe in the official CDK Slack community?

Yes, Slack is a good idea. @pahud has a list of people as well.

**Comment by mmuller88** [comment]
Yep, the construct looks pretty much what I usually do/need to implement when I need the UserPoolClient secret. LGTM
Thanks for the work :)

**Comment by Dzhuneyt** [comment]
@misterjoshua given the positive feedback by @mmuller88, do you feel confident enough to in the changes to approve them or you would prefer further feedback from other members of the CDK team as well?

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (jogold) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:452]:
```suggestion
            resources: [this.userPool.userPoolArn],
```

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (jogold) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:353]: Would call this `_userPoolClientSecret`

**Comment by misterjoshua** [comment]
> @misterjoshua given the positive feedback by @mmuller88, do you feel confident enough in the changes to approve them or you would prefer further feedback from other members of the CDK team as well?

@Dzhuneyt I like @jogold's change requests. Once you've completed these, I'll be happy to approve the PR.

**Review by misterjoshua — APPROVED** [review]
LGTM

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (Dzhuneyt) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts]: @kylelaker good suggestion. Implemented.

**Comment by Dzhuneyt** [comment]
@misterjoshua now that I've addressed also the latest suggestions by @kylelaker, I think it's all in your hands.

**Comment by misterjoshua** [comment]
> @misterjoshua now that I've addressed also the latest suggestions by @kylelaker, and the build succeeding, I think it's now all in your hands.

We'll need to wait for the CDK team to approve it now. :)

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (Dzhuneyt) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts]: IAM policy restricted in scope as per the suggestion.

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (Dzhuneyt) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts]: Renamed.

**Comment by jogold** [comment]
@Dzhuneyt sorry for the bad experience here but I'm not part of the CDK team (https://github.com/aws/aws-cdk/issues/7225#issuecomment-1197213356).

In the meantime, note that you don't have to keep merging from master, the mergify bot will do that once this PR gets approved.

**Comment by TheRealAmazonKendra** [comment]
> > @misterjoshua now that I've addressed also the latest suggestions by @kylelaker, and the build succeeding, I think it's now all in your hands.
> 
> We'll need to wait for the CDK team to approve it now. :)

You called?

I'll do a review of this within the next week. In the meantime, please make sure that your PR title confirms to the conventional commit standard (fix, feat, chore) and that it is written in a style that will reflect correctly in the change log (See [Contributing Guide, Pull Requests](https://github.com/aws/aws-cdk/blob/v1-main/CONTRIBUTING.md#step-4-pull-request)).

Additionally, please make sure that your PR body describes the problem the PR is solving, and the design approach and alternatives considered. Explain why the PR solves the problem. A link to an issue is helpful, but does not replace an explanation of your thought process.

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:426]: Maybe I'm missing something here, but I don't think this can be undefined. You're either generating the `_userPoolClientSecret`, returning it as is if it already exists, or throwing an error.

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:351]: This seems redundant. With a if statement false and undefined will both return as false. But it could also just be written as `private _generateSecret?: boolean;` which is more stylistically consistent.

File: packages/@aws-cdk/aws-cognito/test/user-pool-client.test.ts
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-cognito/test/user-pool-client.test.ts:46]: Why are you using `test.only` here? This means that no other tests in this package will run. We very much do not want that.
```suggestion
  test('client with secret', () => {
```

File: packages/@aws-cdk/aws-cognito/test/user-pool-client.test.ts
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-cognito/test/user-pool-client.test.ts:52]: Each of these should be their own test instead of three cases within one test. Please separate them out.

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
Just a few comments below. Also, just another reminder to make sure the PR title is written according to our guidelines.

**Review by TheRealAmazonKendra — DISMISSED** [review]
Putting this back in changes requested. If you want to merge from main, please use `@Mergifyio update` because it doesn't dismiss the review. This way we can more easily see what needs another look and what is still awaiting changes.

**Comment by Dzhuneyt** [comment]
@Mergifyio update

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (Dzhuneyt) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts]: Agreed. Good suggestion. Implemented.

File: packages/@aws-cdk/aws-cognito/test/user-pool-client.test.ts
Comment (Dzhuneyt) [reply packages/@aws-cdk/aws-cognito/test/user-pool-client.test.ts]: Gotcha. Fixed.

File: packages/@aws-cdk/aws-cognito/test/user-pool-client.test.ts
Comment (Dzhuneyt) [reply packages/@aws-cdk/aws-cognito/test/user-pool-client.test.ts]: That was not intentional. Leftover from local testing. Corrected.

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (Dzhuneyt) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts]: Agreed. Fixed

**Comment by Dzhuneyt** [comment]
@TheRealAmazonKendra this should be ready for another set of eyes now. Just FYI. Thanks for the good tips above.

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
Just a few nitpicky things and then this will be ready to go!

File: packages/@aws-cdk/aws-cognito/test/user-pool-client.test.ts
Comment (Dzhuneyt) [reply packages/@aws-cdk/aws-cognito/test/user-pool-client.test.ts:59]: We are doing it literally one line below.

**Comment by Dzhuneyt** [comment]
@Mergifyio update

**Comment by Dzhuneyt** [comment]
> 

Ready for another round. Applied your suggestions above.

File: packages/@aws-cdk/aws-cognito/test/user-pool-client.test.ts
Comment (TheRealAmazonKendra) [reply packages/@aws-cdk/aws-cognito/test/user-pool-client.test.ts:59]: Not really. They test is testing that the output, whatever it is, is the same when the function is called multiple times. I wanted this test to test against the expected values we should see in the template. While you could argue that the integration tests are doing that, it's within a much larger context and there's a lot more room to miss that the output isn't what we expected.

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-cognito/test/user-pool-client.test.ts
Comment (Dzhuneyt) [reply packages/@aws-cdk/aws-cognito/test/user-pool-client.test.ts:59]: @TheRealAmazonKendra the final CloudFormation template looks like this:
```
      {
        "Type": "AWS::Cognito::UserPoolClient",
        "Properties": {
          "UserPoolId": {
            "Ref": "PoolD3F588B8"
          },
          "AllowedOAuthFlows": [
            "implicit",
            "code"
          ],
          "AllowedOAuthFlowsUserPoolClient": true,
          "AllowedOAuthScopes": [
            "profile",
            "phone",
            "email",
            "openid",
            "aws.cognito.signin.user.admin"
          ],
          "CallbackURLs": [
            "https://example.com"
          ],
          "GenerateSecret": true,
          "SupportedIdentityProviders": [
            "COGNITO"
          ]
        }
      }

```

It doesn't include the generated Client secret, since that one is generated by AWS *after* the Client is created by CloudFormation. So there is no way to retrieve its value in a unit test context.

I guess you were thinking of code resembling the one below, but we simply have nothing to put instead of the `xxx` in that code:
```ts
Template.fromStack(stack).hasResourceProperties('AWS::Cognito::UserPoolClient', {
        xxx: Match.exact(clientWithSecret.userPoolClientSecret),
      });
```

**Comment by Dzhuneyt** [comment]
@TheRealAmazonKendra replied to the thread in https://github.com/aws/aws-cdk/pull/21262#discussion_r935599778

File: packages/@aws-cdk/aws-cognito/test/user-pool-client.test.ts
Comment (TheRealAmazonKendra) [reply packages/@aws-cdk/aws-cognito/test/user-pool-client.test.ts:59]: That's fine. What I want to see tested against is what we put into the template on synth. This is to make sure that (a) the new behavior works as intended and (b) it is not accidentally undone by future changes.

Comment (Dzhuneyt) [reply packages/@aws-cdk/aws-cognito/test/user-pool-client.test.ts:59]: I see. You are referring to tests that check for the existence and correctness of the "GenerateSecret" field in the generated CloudFormaton template. I've now created tests to check for this. Back to you... :)

File: packages/@aws-cdk/aws-cognito/test/user-pool-client.test.ts
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-cognito/test/user-pool-client.test.ts:68]: This is already tested elsewhere in this file. I was meaning the resources that should be generated in this case but not the others. Like this:

```suggestion
      // Make sure the generated template has correct resources
      Template.fromStack(stack).hasResourceProperties('Custom::DescribeCognitoUserPoolClient', {
        ServiceToken: {
          'Fn::GetAtt': [
            'AWS679f53fac002430cb0da5b7982bd22872D164C4C',
            'Arn',
          ],
        },
        Create: {
          'Fn::Join': [
            '',
            [
              '{"region":"',
              {
                Ref: 'AWS::Region',
              },
              '","service":"CognitoIdentityServiceProvider","action":"describeUserPoolClient","parameters":{"UserPoolId":"',
              {
                Ref: 'PoolD3F588B8',
              },
              '","ClientId":"',
              {
                Ref: 'clientWithSecretD25031A8',
              },
              '"},"physicalResourceId":{"id":"',
              {
                Ref: 'clientWithSecretD25031A8',
              },
              '"}}',
            ],
          ],
        },
        InstallLatestAwsSdk: true,
      });

      Template.fromStack(stack).hasResourceProperties('AWS::IAM::Policy', {
        PolicyDocument: {
          Statement: [{
            Action: 'cognito-idp:DescribeUserPoolClient',
            Effect: 'Allow',
            Resource: {
              'Fn::GetAtt': [
                'PoolD3F588B8',
                'Arn',
              ],
            },
          }],
          Version: '2012-10-17',
        },
        PolicyName: 'clientWithSecretDescribeCognitoUserPoolClientCustomResourcePolicyCDE4AB00',
        Roles: [{ Ref: 'AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2' }],
      });

      Template.fromStack(stack).hasResourceProperties('AWS::IAM::Role', {
        AssumeRolePolicyDocument: {
          Statement: [{
            Action: 'sts:AssumeRole',
            Effect: 'Allow',
            Principal: {
              Service: 'lambda.amazonaws.com',
            },
          }],
          Version: '2012-10-17',
        },
        ManagedPolicyArns: [{
          'Fn::Join': [
            '',
            [
              'arn:',
              {
                Ref: 'AWS::Partition',
              },
              ':iam::aws:policy/service-role/AWSLambdaBasicExecutionRole',
            ],
          ],
        }],
      });

      Template.fromStack(stack).hasResourceProperties('AWS::Lambda::Function', {
        Code: {
          S3Bucket: {
            'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
          },
          S3Key: '105b4f39ae68785e705640aa91919e412fcba2dd454aca53412747be8d955286.zip',
        },
        Role: {
          'Fn::GetAtt': [
            'AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2',
            'Arn',
          ],
        },
        Handler: 'index.handler',
        Runtime: 'nodejs14.x',
        Timeout: 120,
      });
    });
```

File: packages/@aws-cdk/aws-cognito/test/user-pool-client.test.ts
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-cognito/test/user-pool-client.test.ts:87]: Versus this:
```suggestion
      // Make sure the generated template does not create resources
      expect(Template.fromStack(stack).findResources('Custom::DescribeCognitoUserPoolClient')).toEqual({});
      expect(Template.fromStack(stack).findResources('AWS::IAM::Policy')).toEqual({});
      expect(Template.fromStack(stack).findResources('AWS::IAM::Role')).toEqual({});
      expect(Template.fromStack(stack).findResources('AWS::Lambda::Function')).toEqual({});
```

File: packages/@aws-cdk/aws-cognito/test/user-pool-client.test.ts
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-cognito/test/user-pool-client.test.ts:107]: And this:
```suggestion
      // Make sure the generated template does not create resources
      expect(Template.fromStack(stack).findResources('Custom::DescribeCognitoUserPoolClient')).toEqual({});
      expect(Template.fromStack(stack).findResources('AWS::IAM::Policy')).toEqual({});
      expect(Template.fromStack(stack).findResources('AWS::IAM::Role')).toEqual({});
      expect(Template.fromStack(stack).findResources('AWS::Lambda::Function')).toEqual({});
```

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
I've left suggestions below with the kind of testing I'm looking for. Feel free to just commit my suggestions directly. I've tested them locally and it passes.

**Review by TheRealAmazonKendra — APPROVED** [review]

**Pull request merged**
